### PR TITLE
[Codegen] Support broadcast op with symbolic shape

### PR DESCRIPTION
**Pull request by yzhliu**
Per https://discuss.tvm.ai/t/rfc-codegen-broadcast-ops-with-symbolic-shape/2908

@tqchen @reminisce @junrushao1994  Please review.

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks for the RFC, I like the idea. Sorry that I didn't get enough time to comment specifically about the behavior, can you please open an RFC issue? So we can discuss it a bit more.

The things to be improved are
- Document the behavior independent of arg_binder.
    - Maps buffer[i][j][k] -> buffer[i][0][k] if dimension i's shape equals 0
- Debate on the name (auto broadcast?), enum vs string as type key
- Discuss how would the behavior affect topi implementation of broadcast and reduce ops
- Think about alternative implementations
  - Do we have to introduce stride?
  -  e.g. j -> min(j, shape[ndim])
  - j -> j * scale(scale = 0 or 1)

**Comment by junrushao1994** [comment]
>   * Do we have to introduce stride?
>   * e.g. j -> min(j, shape[ndim])
>   * j -> j * scale(scale = 0 or 1)

I do think it is proper to introduce `stride`. First, as far as we could see, using strides does not seem to have too much overhead, compared with other workarounds. Second, it seems to pave ways for other possible hacks to the buffer.

**Comment by junrushao1994** [comment]
@were Could you review this PR?

**Comment by were** [comment]
LGTM, too.

**Comment by yzhliu** [comment]
@tqchen modified per suggestion.

File: python/tvm/api.py [hunk 1]
```diff
@@ -574,11 +575,40 @@ def decl_buffer(shape,
         If 0 is pssed, the alignment will be set to 1.
         if non-zero is passed, we will created a Var for elem_offset if elem_offset is not None.
 
+    buffer_type: str, optional, {"", "auto_broadcast"}
+        auto_broadcast buffer allows one to implement broadcast computation
+        without considering whether dimension size equals to one.
+        TVM will insert `strides[i] = shape[i] == 1 ? 0 : strides[i]` during arg binding.
+        See src/pass/arg_binder.cc for reference.
```
Comment (tqchen) [inline python/tvm/api.py]: Let us avoid refer to c++ code in the python side, would be great if python doc can stand by itself

Comment (yzhliu) [reply python/tvm/api.py]: addressed

**Comment by yzhliu** [comment]
Thanks @tqchen @junrushao1994 @were for review.

**Pull request merged**